src/formats/MOL2.cpp

```diff
@@ -0,0 +1,298 @@
+// Chemfiles, a modern library for chemistry file reading and writing
+// Copyright (C) Guillaume Fraux and contributors -- BSD license
+
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include "chemfiles/formats/MOL2.hpp"
+
+#include "chemfiles/ErrorFmt.hpp"
+#include "chemfiles/File.hpp"
+#include "chemfiles/Frame.hpp"
+#include "chemfiles/utils.hpp"
+#include "chemfiles/warnings.hpp"
+
+using namespace chemfiles;
+template<> FormatInfo chemfiles::format_information<MOL2Format>() {
+    return FormatInfo("MOL2").with_extension(".mol2").description(
+        "Tripos mol2 text format"
+    );
+}
+
+/// Fast-forward the file until the tag is found.
+static std::streampos read_until(TextFile& file, const std::string& tag);
+/// Fast-forward the file for one step, returning a valid positionif the file does
+/// contain one more step or -1 if it does not.
+static std::streampos forward(TextFile& file);
+
+MOL2Format::MOL2Format(const std::string& path, File::Mode mode)
+  : file_(TextFile::create(path, mode)) {
+    while (!file_->eof()) {
+        auto position = forward(*file_);
+        if (position == std::streampos(-1)) {
+            break;
+        }
+        if (!file_) {
+            throw format_error("IO error while reading '{}' as MOL2", path);
+        }
+        steps_positions_.push_back(position);
+    }
+    file_->rewind();
+}
+
+size_t MOL2Format::nsteps() {
+    return steps_positions_.size();
+}
+
+void MOL2Format::read_step(const size_t step, Frame& frame) {
+    assert(step < steps_positions_.size());
+    file_->seekg(steps_positions_[step]);
+    read(frame);
+}
+
+void MOL2Format::read(Frame& frame) {
+    const auto& startline= file_->readline();
+    assert(startline.substr(0, 17) == "@<TRIPOS>MOLECULE");
+
+    frame.set("name", trim(file_->readline()) );
+    const auto& line = file_->readline();
+
+    const auto counts = split(line, ' ');
+
+    long long natoms = string2longlong(counts[0]);
+    long long nbonds = 0;
+    
+    if (counts.size() >= 2) {
+        nbonds = string2longlong(counts[1]);
+    }
+
+    residues_.clear();
+    frame.resize(0);
+    frame.reserve(static_cast<size_t>(natoms));
+
+    auto junk = file_->readline();
+    bool charges = (trim(file_->readline()) != "NO_CHARGES");
+
+    while (!file_->eof()) {
+        auto curr_pos = file_->tellg();
+        auto line2 = trim(file_->readline());
```

Comment:
Could this use the same `line` variable ?

Reply:
Yes